File: zeroMQ/cpp/publisher.cpp
#include <zmq.hpp>
#include <string>
#include <iostream>
#include <thread>
#include <chrono>

int main(int argc, char** argv) {
    if (argc < 4) {
        std::cerr << "Usage: " << argv[0] << " <bind-endpoint> <topic> <message>\n";
        std::cerr << "Example: ./publisher tcp://*:5555 test \"hello\"\n";
        return 1;
    }
    std::string endpoint = argv[1];
    std::string topic = argv[2];
    std::string message = argv[3];

    zmq::context_t ctx{1};
    zmq::socket_t pub(ctx, ZMQ_PUB);
    pub.bind(endpoint);

    // allow subscribers to connect
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    std::string payload = topic + " " + message;
    zmq::message_t msg(payload.data(), payload.size());
    try {
        pub.send(msg);
    } catch (const zmq::error_t& e) {
        std::cerr << "Failed to send message: " << e.what() << "\n";
        return 1;
    }
    std::cout << "Published on topic '" << topic << "': " << message << "\n";
    return 0;
}

File: zeroMQ/python/publisher.py
#!/usr/bin/env python3
"""ZeroMQ Python publisher example
Requires: pip install pyzmq
"""
import sys
import zmq

def main():
    if len(sys.argv) < 4:
        print(f"Usage: {sys.argv[0]} <bind-endpoint> <topic> <message>")
        print("Example: ./publisher.py tcp://*:5555 test hello")
        return 1
    endpoint = sys.argv[1]
    topic = sys.argv[2]
    message = sys.argv[3]

    ctx = zmq.Context()
    pub = ctx.socket(zmq.PUB)
    pub.bind(endpoint)

    # allow subscribers to connect
    import time; time.sleep(0.2)

    pub.send_string(f"{topic} {message}")
    print(f"Published on {topic}: {message}")
    return 0

if __name__ == '__main__':
    sys.exit(main())
